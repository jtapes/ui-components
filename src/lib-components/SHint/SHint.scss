@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-hint {
  position: relative;
  display: inline-flex;

  &__inner {
    display: inline-flex;
    align-items: center;
    justify-content: center;
    border: 1px solid $color--border;
    border-radius: 50%;
    width: $size--28;
    height: $size--28;
    color: $color--font-secondary;
    cursor: pointer;
  }

  &__wrapper {
    position: absolute;
    display: inline-block;
    box-shadow: 0 4px 40px $color--shadow;
    background-color: $color--white;
    transition: opacity 0.2s, visibility 0.2s;

    &::after {
      content: '';
      position: absolute;
      width: 0;
      height: 0;
      box-sizing: border-box;
    }

    &--hidden {
      opacity: 0;
      visibility: hidden;
    }
  }

  &__wrapper.s-hint__wrapper--appended-to-body {
    @include z-index(above);

    &.s-hint__wrapper {
      &--top-left {
        &::after {
          right: 0;
          bottom: -18px;
          border-top: 18px solid $color--white;
          border-left: 21px solid transparent;
        }
      }

      &--top-right {
        &::after {
          left: 0;
          bottom: -18px;
          border-top: 18px solid $color--white;
          border-right: 21px solid transparent;
        }
      }

      &--bottom-left {
        &::after {
          right: 0;
          top: -18px;
          border-bottom: 18px solid $color--white;
          border-left: 21px solid transparent;
        }
      }

      &--bottom-right {
        &::after {
          left: 0;
          top: -18px;
          border-right: 21px solid transparent;
          border-bottom: 18px solid $color--white;
        }
      }
    }
  }

  &__wrapper:not(.s-hint__wrapper--appended-to-body) {
    @include z-index(popover);

    &.s-hint__wrapper {
      &--top-left {
        right: 50%;
        bottom: calc(100% + 18px);

        &::after {
          right: 0;
          bottom: -18px;
          border-top: 18px solid $color--white;
          border-left: 21px solid transparent;
        }
      }

      &--top-right {
        left: 50%;
        bottom: calc(100% + 18px);

        &::after {
          left: 0;
          bottom: -18px;
          border-top: 18px solid $color--white;
          border-right: 21px solid transparent;
        }
      }

      &--bottom-left {
        right: 50%;
        top: calc(100% + 18px);

        &::after {
          right: 0;
          top: -18px;
          border-bottom: 18px solid $color--white;
          border-left: 21px solid transparent;
        }
      }

      &--bottom-right {
        left: 50%;
        top: calc(100% + 18px);

        &::after {
          left: 0;
          top: -18px;
          border-right: 21px solid transparent;
          border-bottom: 18px solid $color--white;
        }
      }
    }
  }

  &__content {
    display: inline-block;
    padding: 0 $size--20 $size--16 $size--20;
  }

  &__header {
    display: flex;
  }

  &__close {
    display: flex;
    flex-grow: 1;
    align-self: self-start;
    justify-content: flex-end;
    padding: $size--4;
    color: $color--primary;
    background-color: transparent;
    cursor: pointer;
  }
}
