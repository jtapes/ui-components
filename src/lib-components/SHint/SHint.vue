<template>
  <div
    class="s-hint"
    :class="{ 'mx-8': !$slots.default }"
    @mouseleave="handleVisibleChange(false)"
    @mouseenter="handleVisibleChange(true)"
  >
    <!-- @slot Слот для компонента, на котором висит hint  -->
    <slot>
      <span class="s-hint__inner">
        <svg-icon name="question-16" class="icon-16" />
      </span>
    </slot>
    <div
      v-show="!disabled"
      ref="hint"
      class="s-hint__wrapper"
      :class="[
        `s-hint__wrapper--${placement}`,
        { 's-hint__wrapper--appended-to-body': appendToBody },
        { 's-hint__wrapper--hidden': !hintVisible },
      ]"
      :style="{ minWidth, maxWidth, ...appendedToBodyPosition, ...appendedToBodySize }"
      @mouseleave="handleVisibleChange(false)"
      @mouseenter="handleVisibleChange(true)"
    >
      <div class="s-hint__header">
        <!--  @slot Слот контента хедера -->
        <slot name="header" />

        <button v-if="closable" class="s-hint__close" @click="handleCloseHint">
          <!-- @slot Слот иконки закрытия -->
          <slot name="close-icon">
            <svg-icon class="icon-16" name="close-24-32" />
          </slot>
        </button>
      </div>
      <div class="s-hint__content text-sm" :class="{ 'pt-16': !closable }">
        <!--  @slot  Слот контента хинта  -->
        <slot name="content">
          {{ content }}
        </slot>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import { Vue, Component, Prop, Model } from 'vue-property-decorator'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import SButton from '@/lib-components/SButton/SButton.vue'

type AppendedToBodyPosition = {
  top?: string
  left?: string
}

type AppendedToBodySize = {
  minHeight?: string
  maxWidth?: string
}

const PLACEMENT = ['top-left', 'top-right', 'bottom-left', 'bottom-right']

/** Компонент для отображения подсказок  */
@Component({
  components: {
    SButton,
    SvgIcon,
  }
})
export default class SHint extends Vue {
  visible = false
  appendedToBodyPosition: AppendedToBodyPosition = {}
  appendedToBodySize: AppendedToBodySize = {}

  /** Текст хинта, может быть заменен слотом #content */
  @Prop(String)
  content?: string

  /** Свойство отвечающее за дефолтное поведение hint ( hover ) */
  @Prop({ type: Boolean, default: true })
  trigger!: boolean

  /** Положение hint ('top-left', 'top-right', 'bottom-left', 'bottom-right') */
  @Prop({
    type: String,
    default: 'top-left',
    validator(value: string): boolean {
      return PLACEMENT.includes(value)
    },
  })
  placement!: string

  /** Минимальная длина контента hint */
  @Prop(String)
  minWidth?: string

  /** Максимальная длина контента hint */
  @Prop(String)
  maxWidth?: string

  /** Свойство отключения хинта */
  @Prop({ type: Boolean, default: false })
  disabled!: boolean

  /** Свойство выноса хинта в body документа */
  @Prop({ type: Boolean, default: false })
  appendToBody!: boolean

  /** Свойство добавления кнопки закрытия хинта */
  @Prop({ type: Boolean, default: false })
  closable!: boolean

  @Model('change')
  readonly value?: boolean

  get hintVisible(): boolean {
    const visible = this.value !== undefined ? this.value : this.visible
    if (visible && this.appendToBody) {
      this.prepareHint()
    }

    return visible
  }

  set hintVisible(value: boolean) {
    if (this.value !== undefined) {
      this.$emit('change', value)
    } else {
      this.visible = value
    }
  }

  handleVisibleChange(visible: boolean): void {
    if (this.trigger) {
      this.hintVisible = visible
    }
  }

  handleCloseHint(): void {
    this.hintVisible = false
  }

  prepareHint(): void {
    if (!this.$isServer) {
      this.$nextTick(async () => {
        if (this.$slots.default && this.$slots.default[0].elm && this.$refs.hint) {
          const slotRect = (this.$slots.default[0].elm as Element).getBoundingClientRect()
          const hintRect = (this.$refs.hint as Element).getBoundingClientRect()
          let hintWidth
          if (this.minWidth && parseInt(this.minWidth) > slotRect.width) {
            hintWidth = parseInt(this.minWidth)
          } else if (this.maxWidth) {
            hintWidth = parseInt(this.maxWidth)
          } else {
            hintWidth = slotRect.width
          }
          await this.calculateSize(slotRect, hintWidth)
          await this.calculatePosition(slotRect, hintRect, hintWidth)
        }
      })
    }
  }

  calculateSize(slotRect: DOMRect, hintWidth: number): void {
    this.appendedToBodySize = {
      minHeight: `${slotRect.height}px`,
      maxWidth: `${hintWidth}px`,
    }
  }

  calculatePosition(slotRect: DOMRect, hintRect: DOMRect, hintWidth: number): void {
    switch (this.placement) {
      case 'top-left': {
        this.appendedToBodyPosition = {
          top: `${slotRect.top - hintRect.height - 18 + window.scrollY}px`,
          left: `${slotRect.left - hintWidth + slotRect.width / 2}px`,
        }
        break
      }
      case 'top-right': {
        this.appendedToBodyPosition = {
          top: `${slotRect.top - hintRect.height - 18 + window.scrollY}px`,
          left: `${slotRect.left + hintWidth - slotRect.width / 2}px`,
        }
        break
      }
      case 'bottom-left': {
        this.appendedToBodyPosition = {
          top: `${slotRect.top + slotRect.height + 18 + window.scrollY}px`,
          left: `${slotRect.left - hintWidth + slotRect.width / 2}px`,
        }
        break
      }
      case 'bottom-right': {
        this.appendedToBodyPosition = {
          top: `${slotRect.top + slotRect.height + 18 + window.scrollY}px`,
          left: `${slotRect.left + hintWidth - slotRect.width / 2}px`,
        }
        break
      }
    }
  }

  mounted(): void {
    if (this.appendToBody) {
      this.prepareHint()
      document.body.append((this.$refs.hint as Node))
    }
  }
}
</script>
