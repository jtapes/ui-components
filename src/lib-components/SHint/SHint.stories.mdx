import { action } from '@storybook/addon-actions'
import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest';
import {
  Meta,
  Story,
  ArgsTable,
  Description,
  Source,
  Props,
  Preview,
} from '@storybook/addon-docs'
import results from '/.jest-test-results.json';
import SHint from '@/lib-components/SHint/SHint.vue'
import SButton from '@/lib-components/SButton/SButton.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'

<Meta
  title="DESIGN SYSTEM/Atom/Hint"
  component={SHint}
  decorators={[withDesign, withTests({results})]}
  argTypes={{
    content: { control: 'text' },
    trigger: { control: 'boolean', defaultValue: true },
    placement: {
        options: [
            'top-left',
            'top-right',
            'bottom-left',
            'bottom-right',
        ],
        control: { type: 'select' },
        defaultValue: 'top-left',
    },
    minWidth: { control: 'text' },
    closable: { control: 'boolean' }
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/IIeFkxNoHGcaqNHlVeNrhW/MVP-3-Market-UI-Kit?node-id=3449%3A2819',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Подсказка

<Description of={SHint} />

export const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { SHint, SButton, SvgIcon },
  template: `
    <div style="width: 100%; height: 300px; display: flex; align-items: center; justify-content: center;">
      <s-hint :content="content" :trigger="trigger" :placement="placement" :min-width="minWidth" :closable="closable">
        <s-button mode="outline">
          В корзине
          <template #append>
            <svg-icon class="icon-32" name="checked-24-32" />
          </template>
        </s-button>
      </s-hint>
    </div>
  `,
})

<Preview withToolbar withSource="none">
  <Story name="SHint" args={{ content: 'В корзину добавлено 20 т исходя из автоподбора', minWidth: '250px' }}>
    {Template.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`
<s-hint content="В корзину добавлено 20 т исходя из автоподбора">
  <s-button mode="outline">
    В корзине
    <template #append>
      <svg-icon class="icon-32" name="checked-24-32" />
    </template>
  </s-button>
</s-hint>
  `}
/>

<ArgsTable of={SHint} />
