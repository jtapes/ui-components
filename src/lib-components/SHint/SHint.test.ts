import { mount } from '@vue/test-utils'
import SHint from '@/lib-components/SHint/SHint.vue'

describe('SHint', () => {
  const hintContent = 'Подсказка'

  const template = `
    <s-hint content="${hintContent}" v-model="hintVisible">
      <span>Hover</span>
    </s-hint>
  `

  test('Is Vue snapshot', () => {
    const wrapper = mount({
      template,
      components: { SHint },
      data: () => ({
        hintVisible: true,
      }),
    })

    expect(wrapper.element).toMatchSnapshot()
  })
  test('Should change visible with v-model', async () => {
    const wrapper = mount({
      template,
      components: { SHint },
      data: () => ({
        hintVisible: true,
      }),
    })


    await wrapper.trigger('mouseleave')
    await wrapper.vm.$nextTick()

    const hint = wrapper.find('.s-hint__wrapper')

    expect(hint.element.classList.contains('s-hint__wrapper--hidden')).toBe(true)
    expect(wrapper.vm.$data.hintVisible).toBe(false)
  })
  test('Should change visible without v-model', async () => {
    const wrapper = mount({
      template: `
        <s-hint content="${hintContent}">
          <span>Hover</span>
        </s-hint>
      `,
      components: { SHint },
    })
    const hint = wrapper.find('.s-hint__wrapper')

    await wrapper.trigger('mouseenter')
    expect(hint.element.classList.contains('s-hint__wrapper--hidden')).toBe(false)

    await wrapper.trigger('mouseleave')
    expect(hint.element.classList.contains('s-hint__wrapper--hidden')).toBe(true)
  })
  test('Component should has content of content prop', () => {
    const wrapper = mount({
      template,
      components: { SHint },
      data: () => ({
        hintVisible: true,
      }),
    })

    expect(wrapper.find('.s-hint__content').text()).toBe(hintContent)
  })
})
