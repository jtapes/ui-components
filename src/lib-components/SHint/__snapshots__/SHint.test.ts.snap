// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SHint Is Vue snapshot 1`] = `
<div
  class="s-hint"
>
  <span>
    Hover
  </span>
   
  <div
    class="s-hint__wrapper s-hint__wrapper--top-left"
  >
    <div
      class="s-hint__header"
    >
       
      <!---->
    </div>
     
    <div
      class="s-hint__content text-sm pt-16"
    >
      
        Подсказка
      
    </div>
  </div>
</div>
`;
